lib/puppet/resource/type.rb

```diff
@@ -427,7 +427,15 @@ def validate_resource_hash(resource, resource_hash)
   # Validate that all parameters given to the resource are correct
   # @param resource [Puppet::Resource] the resource to validate
   def validate_resource(resource)
-    validate_resource_hash(resource, Hash[resource.parameters.map { |name, value| [name.to_s, value.value] }])
+    # Since Sensitive values have special encoding (in a separate parameter) an unwrapped sensitive value must be
+    # recreated as a Sensitive in order to perform correct type checking.
+    sensitives = Set.new(resource.sensitive_parameters)
+    validate_resource_hash(resource,
+      Hash[resource.parameters.map do |name, value|
+        value_to_validate = sensitives.include?(name) ? Puppet::Pops::Types::PSensitiveType::Sensitive.new(value.value) : value.value
```

Comment:
@branan This is the insane part. The Sensitive values have at this point been unwrapped and marked in `sensitive_parameters`. The type signature for the resource type may however have typed a parameter with `Sensitive`. To make it work an instance of `Sensitive` must be used when checking the type signature - here done as a Hash of values checked against a Struct data type.